panic: runtime error: invalid memory address or nil pointer dereference

panic: runtime error: invalid memory address or nil pointer dereference

I keep getting this runtime error when I try to run my test. 

```
panic: runtime error: invalid memory address or nil pointer dereference

    panic: runtime error: invalid memory address or nil pointer dereference

[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xca727b]
```

The contract compiles and shows signs that it is about to start fuzzing, but it doesn't progress beyond zero sequences (breaks on 1 seq/s). Something like this: `⇾ fuzz: elapsed: 1m23s, calls: 135 (41/sec), seq/s: 1, coverage: 8`. I don't get the error with Echidna and Foundry.

Unfortunately, I can't publicly share the repo but I am confident the issue isn't with the setup.

For additional context, please check out my [Echidna issue submission.](https://github.com/crytic/echidna/issues/1296#issuecomment-2246922806)


Medusa version: 0.1.5
Slither version: 0.10.3

The full error:
<details>
```
⇾ fuzz: elapsed: 1m23s, calls: 135 (41/sec), seq/s: 1, coverage: 8

panic: runtime error: invalid memory address or nil pointer dereference

	panic: runtime error: invalid memory address or nil pointer dereference

[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xca727b]



goroutine 12649 [running]:

github.com/crytic/medusa/fuzzing/executiontracer.(*ExecutionTracer).OnTxEnd(...)

	github.com/crytic/medusa/fuzzing/executiontracer/execution_tracer.go:113

github.com/crytic/medusa/chain/vendored.EVMApplyTransaction.func1()

	github.com/crytic/medusa/chain/vendored/apply_transaction.go:45 +0x35

panic({0xe75d20?, 0x17cb640?})

	runtime/panic.go:770 +0x132

github.com/crytic/medusa/chain.getStandardCheatCodeContract.func13(0xc00e886140, {0xc00217fd60, 0x1, 0x1})

	github.com/crytic/medusa/chain/standard_cheat_code_contract.go:248 +0x57

github.com/crytic/medusa/chain.(*CheatCodeContract).Run(0xc0005f8008, {0xc01e9f5640, 0x24, 0x40})

	github.com/crytic/medusa/chain/cheat_code_contract.go:180 +0x103

github.com/ethereum/go-ethereum/core/vm.RunPrecompiledContract({0x11abd00, 0xc0005f8008}, {0xc01e9f5640, 0x24, 0x40}, 0x1742fa0, 0xc00d2085a0)

	github.com/ethereum/go-ethereum@v1.14.6/core/vm/contracts.go:204 +0xb3

github.com/ethereum/go-ethereum/core/vm.(*EVM).Call(0xc00d571e00, {0x11a9f80, 0xc00cdf5ad0}, {0x71, 0x9, 0x70, 0x9e, 0xcf, 0xa9, 0x1a, ...}, ...)

	github.com/ethereum/go-ethereum@v1.14.6/core/vm/evm.go:233 +0x651

github.com/ethereum/go-ethereum/core/vm.opCall(0x1?, 0xc00d550660, 0xc0102c6f18)

	github.com/ethereum/go-ethereum@v1.14.6/core/vm/instructions.go:683 +0x576

github.com/ethereum/go-ethereum/core/vm.(*EVMInterpreter).Run(0xc00d550660, 0xc00cdf5ad0, {0xc00e003ea0, 0x44, 0x50}, 0x0)

	github.com/ethereum/go-ethereum@v1.14.6/core/vm/interpreter.go:304 +0xc9c

github.com/ethereum/go-ethereum/core/vm.(*EVM).Call(0xc00d571e00, {0x11a9a20, 0xc00d574240}, {0xa6, 0x47, 0xff, 0x3c, 0x36, 0xcf, 0xab, ...}, ...)

	github.com/ethereum/go-ethereum@v1.14.6/core/vm/evm.go:249 +0x948

github.com/ethereum/go-ethereum/core.(*StateTransition).TransitionDb(0xc02770b268)

	github.com/ethereum/go-ethereum@v1.14.6/core/state_transition.go:445 +0x805

github.com/ethereum/go-ethereum/core.ApplyMessage(0x65025c79c4d55a2b?, 0x215e217a7c31f814?, 0xcfd6cc8d?)

	github.com/ethereum/go-ethereum@v1.14.6/core/state_transition.go:184 +0x57

github.com/crytic/medusa/chain/vendored.EVMApplyTransaction(0xc00c4573f0, 0xc00058a820, 0xc0004299f8, 0x57cb2?, 0xc01e8b1fb0, 0xc00bc26340, 0xc00ca6b760, {0x68, 0xbf, 0x11, ...}, ...)

	github.com/crytic/medusa/chain/vendored/apply_transaction.go:54 +0x273

github.com/crytic/medusa/chain.(*TestChain).PendingBlockAddTx(0xc01630a3c0, 0xc00c4573f0, 0x57cb2?)

	github.com/crytic/medusa/chain/test_chain.go:778 +0x48f

github.com/crytic/medusa/fuzzing/calls.ExecuteCallSequenceIteratively(0xc01630a3c0, 0xc02770bb90, 0x0, 0xc02770bbb0)

	github.com/crytic/medusa/fuzzing/calls/call_sequence_execution.go:93 +0x3dd

github.com/crytic/medusa/fuzzing/calls.ExecuteCallSequenceWithExecutionTracer(0xc01630a3c0, {0xc007f35b08, 0xe7, 0x11f}, {0xc0174ace58, 0x1, 0x1}, 0x0)

	github.com/crytic/medusa/fuzzing/calls/call_sequence_execution.go:192 +0x55a

github.com/crytic/medusa/fuzzing.(*PropertyTestCaseProvider).callSequencePostCallTest.func2(0xc0080ec6c0, {0xc0174ace58, 0x1, 0x1}, 0x0)

	github.com/crytic/medusa/fuzzing/test_case_property_provider.go:316 +0xee

github.com/crytic/medusa/fuzzing.(*FuzzerWorker).shrinkCallSequence(0xc0080ec6c0, {0xc012cc7100?, 0xc0080ec6c0?, 0xc01630a3c0?}, {0xc00e8dc8e8?, 0xc00bb22ed0?, 0x2?})

	github.com/crytic/medusa/fuzzing/fuzzer_worker.go:521 +0x18c

github.com/crytic/medusa/fuzzing.(*FuzzerWorker).run(0xc0080ec6c0, 0x11b5640?)

	github.com/crytic/medusa/fuzzing/fuzzer_worker.go:613 +0x5e5

github.com/crytic/medusa/fuzzing.(*Fuzzer).spawnWorkersLoop.func1({0xb43624?, 0xc01189a960?})

	github.com/crytic/medusa/fuzzing/fuzzer.go:640 +0x169

created by github.com/crytic/medusa/fuzzing.(*Fuzzer).spawnWorkersLoop in goroutine 1

	github.com/crytic/medusa/fuzzing/fuzzer.go:623 +0x1d3

```
</details>

--- a/chain/test_chain.go
+++ b/chain/test_chain.go
@@ -9,7 +9,6 @@ import (
 	"github.com/crytic/medusa/chain/config"
 	"github.com/ethereum/go-ethereum/core/rawdb"
 	"github.com/ethereum/go-ethereum/core/tracing"
-	"github.com/ethereum/go-ethereum/eth/tracers"
 	"github.com/ethereum/go-ethereum/triedb"
 	"github.com/ethereum/go-ethereum/triedb/hashdb"
 	"github.com/holiman/uint256"
@@ -256,7 +255,7 @@ func (t *TestChain) Clone(onCreateFunc func(chain *TestChain) error) (*TestChain
 		// Now add each transaction/message to it.
 		messages := t.blocks[i].Messages
 		for j := 0; j < len(messages); j++ {
-			err = targetChain.PendingBlockAddTx(messages[j], nil)
+			err = targetChain.PendingBlockAddTx(messages[j])
 			if err != nil {
 				return nil, err
 			}
@@ -728,15 +727,7 @@ func (t *TestChain) PendingBlockCreateWithParameters(blockNumber uint64, blockTi
 // PendingBlockAddTx takes a message (internal txs) and adds it to the current pending block, updating the header
 // with relevant execution information. If a pending block was not created, an error is returned.
 // Returns an error if one occurred.
-func (t *TestChain) PendingBlockAddTx(message *core.Message, getTracerFn func(txIndex int, txHash common.Hash) *tracers.Tracer) error {
-	// Caller can specify any tracer they wish to use for this transaction
-	if getTracerFn == nil {
-		// TODO: Figure out whether it is possible to identify _which_ transaction you want to trace (versus all)
-		getTracerFn = func(txIndex int, txHash common.Hash) *tracers.Tracer {
-			return t.transactionTracerRouter.NativeTracer().Tracer
-		}
-	}
-
+func (t *TestChain) PendingBlockAddTx(message *core.Message, additionalTracers ...*TestChainTracer) error {
 	// If we don't have a pending block, return an error
 	if t.pendingBlock == nil {
 		return errors.New("could not add tx to the chain's pending block because no pending block was created")
@@ -757,11 +748,19 @@ func (t *TestChain) PendingBlockAddTx(message *core.Message, getTracerFn func(tx
 		ConfigExtensions: t.vmConfigExtensions,
 	}
 
-	// Set the tracer to be used in the vm config
-	tracer := getTracerFn(len(t.pendingBlock.Messages), tx.Hash())
-	if tracer != nil {
-		vmConfig.Tracer = tracer.Hooks
+	// Figure out whether we need to attach any more tracers
+	var extendedTracerRouter *TestChainTracerRouter
+	if len(additionalTracers) > 0 {
+		// If we have more tracers, extend the transaction tracer router's tracers with additional ones
+		extendedTracerRouter = NewTestChainTracerRouter()
+		extendedTracerRouter.AddTracer(t.transactionTracerRouter.NativeTracer())
+		t.transactionTracerRouter.AddTracers(additionalTracers...)
+	} else {
+		extendedTracerRouter = t.transactionTracerRouter
 	}
+
+	// Update the VM's tracer
+	vmConfig.Tracer = extendedTracerRouter.NativeTracer().Tracer.Hooks
 
 	// Set tx context
 	t.state.SetTxContext(tx.Hash(), len(t.pendingBlock.Messages))

--- a/chain/test_chain.go
+++ b/chain/test_chain.go
@@ -754,7 +754,7 @@ func (t *TestChain) PendingBlockAddTx(message *core.Message, additionalTracers .
 		// If we have more tracers, extend the transaction tracer router's tracers with additional ones
 		extendedTracerRouter = NewTestChainTracerRouter()
 		extendedTracerRouter.AddTracer(t.transactionTracerRouter.NativeTracer())
-		t.transactionTracerRouter.AddTracers(additionalTracers...)
+		extendedTracerRouter.AddTracers(additionalTracers...)
 	} else {
 		extendedTracerRouter = t.transactionTracerRouter
 	}

--- a/chain/test_chain_test.go
+++ b/chain/test_chain_test.go
@@ -260,7 +260,7 @@ func TestChainDynamicDeployments(t *testing.T) {
 						assert.NoError(t, err)
 
 						// Add our transaction to the block
-						err = chain.PendingBlockAddTx(&msg, nil)
+						err = chain.PendingBlockAddTx(&msg)
 						assert.NoError(t, err)
 
 						// Commit the pending block to the chain, so it becomes the new head.
@@ -385,7 +385,7 @@ func TestChainDeploymentWithArgs(t *testing.T) {
 					assert.NoError(t, err)
 
 					// Add our transaction to the block
-					err = chain.PendingBlockAddTx(&msg, nil)
+					err = chain.PendingBlockAddTx(&msg)
 					assert.NoError(t, err)
 
 					// Commit the pending block to the chain, so it becomes the new head.
@@ -494,7 +494,7 @@ func TestChainCloning(t *testing.T) {
 							assert.NoError(t, err)
 
 							// Add our transaction to the block
-							err = chain.PendingBlockAddTx(&msg, nil)
+							err = chain.PendingBlockAddTx(&msg)
 							assert.NoError(t, err)
 
 							// Commit the pending block to the chain, so it becomes the new head.
@@ -588,7 +588,7 @@ func TestChainCallSequenceReplayMatchSimple(t *testing.T) {
 							assert.NoError(t, err)
 
 							// Add our transaction to the block
-							err = chain.PendingBlockAddTx(&msg, nil)
+							err = chain.PendingBlockAddTx(&msg)
 							assert.NoError(t, err)
 
 							// Commit the pending block to the chain, so it becomes the new head.
@@ -627,7 +627,7 @@ func TestChainCallSequenceReplayMatchSimple(t *testing.T) {
 			_, err := recreatedChain.PendingBlockCreate()
 			assert.NoError(t, err)
 			for _, message := range chain.blocks[i].Messages {
-				err = recreatedChain.PendingBlockAddTx(message, nil)
+				err = recreatedChain.PendingBlockAddTx(message)
 				assert.NoError(t, err)
 			}
 			err = recreatedChain.PendingBlockCommit()

--- a/fuzzing/calls/call_sequence_execution.go
+++ b/fuzzing/calls/call_sequence_execution.go
@@ -7,8 +7,6 @@ import (
 	"github.com/crytic/medusa/fuzzing/contracts"
 	"github.com/crytic/medusa/fuzzing/executiontracer"
 	"github.com/crytic/medusa/utils"
-	"github.com/ethereum/go-ethereum/common"
-	"github.com/ethereum/go-ethereum/eth/tracers"
 )
 
 // ExecuteCallSequenceFetchElementFunc describes a function that is called to obtain the next call sequence element to
@@ -28,7 +26,7 @@ type ExecuteCallSequenceExecutionCheckFunc func(currentExecutedSequence CallSequ
 // A "post element executed check" function is provided to check whether execution should stop after each element is
 // executed.
 // Returns the call sequence which was executed and an error if one occurs.
-func ExecuteCallSequenceIteratively(chain *chain.TestChain, fetchElementFunc ExecuteCallSequenceFetchElementFunc, executionCheckFunc ExecuteCallSequenceExecutionCheckFunc, getTracerFn func(txIndex int, txHash common.Hash) *tracers.Tracer) (CallSequence, error) {
+func ExecuteCallSequenceIteratively(chain *chain.TestChain, fetchElementFunc ExecuteCallSequenceFetchElementFunc, executionCheckFunc ExecuteCallSequenceExecutionCheckFunc, additionalTracers ...*chain.TestChainTracer) (CallSequence, error) {
 	// If there is no fetch element function provided, throw an error
 	if fetchElementFunc == nil {
 		return nil, fmt.Errorf("could not execute call sequence on chain as the 'fetch element function' provided was nil")
@@ -90,7 +88,7 @@ func ExecuteCallSequenceIteratively(chain *chain.TestChain, fetchElementFunc Exe
 			}
 
 			// Try to add our transaction to this block.
-			err = chain.PendingBlockAddTx(callSequenceElement.Call.ToCoreMessage(), getTracerFn)
+			err = chain.PendingBlockAddTx(callSequenceElement.Call.ToCoreMessage(), additionalTracers...)
 
 			if err != nil {
 				// If we encountered a block gas limit error, this tx is too expensive to fit in this block.
@@ -168,7 +166,7 @@ func ExecuteCallSequence(chain *chain.TestChain, callSequence CallSequence) (Cal
 		return nil, nil
 	}
 
-	return ExecuteCallSequenceIteratively(chain, fetchElementFunc, nil, nil)
+	return ExecuteCallSequenceIteratively(chain, fetchElementFunc, nil)
 }
 
 // ExecuteCallSequenceWithTracer attaches an executiontracer.ExecutionTracer to ExecuteCallSequenceIteratively and attaches execution traces to the call sequence elements.
@@ -176,9 +174,6 @@ func ExecuteCallSequenceWithExecutionTracer(testChain *chain.TestChain, contract
 	// Create a new execution tracer
 	executionTracer := executiontracer.NewExecutionTracer(contractDefinitions, testChain.CheatCodeContracts())
 	defer executionTracer.Close()
-	getTracerFunc := func(txIndex int, txHash common.Hash) *tracers.Tracer {
-		return executionTracer.NativeTracer().Tracer
-	}
 
 	// Execute our sequence with a simple fetch operation provided to obtain each element.
 	fetchElementFunc := func(currentIndex int) (*CallSequenceElement, error) {
@@ -188,8 +183,8 @@ func ExecuteCallSequenceWithExecutionTracer(testChain *chain.TestChain, contract
 		return nil, nil
 	}
 
-	// Execute the call sequence
-	executedCallSeq, err := ExecuteCallSequenceIteratively(testChain, fetchElementFunc, nil, getTracerFunc)
+	// Execute the call sequence and attach the execution tracer
+	executedCallSeq, err := ExecuteCallSequenceIteratively(testChain, fetchElementFunc, nil, executionTracer.NativeTracer())
 
 	// By default, we only trace the last element in the call sequence.
 	traceFrom := len(callSequence) - 1

--- a/fuzzing/corpus/corpus.go
+++ b/fuzzing/corpus/corpus.go
@@ -210,7 +210,7 @@ func (c *Corpus) initializeSequences(sequenceFiles *corpusDirectory[calls.CallSe
 		}
 
 		// Execute each call sequence, populating runtime data and collecting coverage data along the way.
-		_, err = calls.ExecuteCallSequenceIteratively(testChain, fetchElementFunc, executionCheckFunc, nil)
+		_, err = calls.ExecuteCallSequenceIteratively(testChain, fetchElementFunc, executionCheckFunc)
 
 		// If we failed to replay a sequence and measure coverage due to an unexpected error, report it.
 		if err != nil {

--- a/fuzzing/executiontracer/execution_tracer.go
+++ b/fuzzing/executiontracer/execution_tracer.go
@@ -119,6 +119,8 @@ func (t *ExecutionTracer) OnTxStart(vm *tracing.VMContext, tx *coretypes.Transac
 	t.trace = newExecutionTrace(t.contractDefinitions)
 	t.currentCallFrame = nil
 	t.onNextCaptureState = nil
+	t.traceMap = make(map[common.Hash]*ExecutionTrace)
+
 	// Store our evm reference
 	t.evmContext = vm
 }

--- a/fuzzing/fuzzer.go
+++ b/fuzzing/fuzzer.go
@@ -463,7 +463,7 @@ func chainSetupFromCompilations(fuzzer *Fuzzer, testChain *chain.TestChain) (*ex
 				}
 
 				// Add our transaction to the block
-				err = testChain.PendingBlockAddTx(msg.ToCoreMessage(), nil)
+				err = testChain.PendingBlockAddTx(msg.ToCoreMessage())
 				if err != nil {
 					return nil, err
 				}

--- a/fuzzing/fuzzer_worker.go
+++ b/fuzzing/fuzzer_worker.go
@@ -317,7 +317,7 @@ func (fw *FuzzerWorker) testNextCallSequence() (calls.CallSequence, []ShrinkCall
 	}
 
 	// Execute our call sequence.
-	testedCallSequence, err := calls.ExecuteCallSequenceIteratively(fw.chain, fetchElementFunc, executionCheckFunc, nil)
+	testedCallSequence, err := calls.ExecuteCallSequenceIteratively(fw.chain, fetchElementFunc, executionCheckFunc)
 
 	// If we encountered an error, report it.
 	if err != nil {
@@ -383,7 +383,7 @@ func (fw *FuzzerWorker) testShrunkenCallSequence(possibleShrunkSequence calls.Ca
 	}
 
 	// Execute our call sequence.
-	_, err = calls.ExecuteCallSequenceIteratively(fw.chain, fetchElementFunc, executionCheckFunc, nil)
+	_, err = calls.ExecuteCallSequenceIteratively(fw.chain, fetchElementFunc, executionCheckFunc)
 	if err != nil {
 		return false, err
 	}

